File: c++/leetcode/1008. Construct Binary Search Tree from Preorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
  TreeNode *bst(vector<int> &, int, int &);
  TreeNode *bstFromPreorder(vector<int> &preorder)
  {
    if (!preorder.size())
      return nullptr;
    auto ind = 0;
    return bst(preorder, INT_MAX, ind);
  }
};

TreeNode *Solution::bst(vector<int> &preorder, int upper, int &ind)
{
  if (ind == preorder.size() || preorder[ind] > upper)
    return nullptr;

  auto root = new TreeNode(preorder[ind]);
  ++ind;
  root->left = bst(preorder, root->val, ind);
  root->right = bst(preorder, upper, ind);

  return root;
}

File: javascript/leetcode/1008. Construct Binary Search Tree from Preorder Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} preorder
 * @return {TreeNode}
 */
var bstFromPreorder = function (preorder) {
  let ind = 0;

  const bst = (preorder, upper) => {
    if (ind === preorder.length || preorder[ind] > upper) return null;

    const root = new TreeNode(preorder[ind]);
    ++ind;
    root.left = bst(preorder, root.val);
    root.right = bst(preorder, upper);
    return root;
  };

  return bst(preorder, Number.MAX_VALUE);
};
